druntime/src/core/demangle.d

```diff
@@ -2924,72 +2903,141 @@ private struct Buffer
-    void remove(scope const(char)[] val) scope
+    void remove(scope BufSlice val) scope
     {
         version (DigitalMars) pragma(inline, false); // tame dmd inliner
 
         if ( val.length )
         {
             assert( contains( dst[0 .. len], val ) );
-            debug(info) printf( "removing (%.*s)\n", cast(int) val.length, val.ptr );
-            size_t v = &val[0] - &dst[0];
+
             assert( len >= val.length && len <= dst.length );
             len -= val.length;
-            for (size_t p = v; p < len; p++)
+            for (size_t p = val.from; p < len; p++)
                 dst[p] = dst[p + val.length];
         }
     }
 
-    char[] append(scope const(char)[] val) return scope
+    void append(scope const(char)[] val) scope
     {
         version (DigitalMars) pragma(inline, false); // tame dmd inliner
 
         if (val.length)
         {
             if ( !dst.length )
                 dst.length = minSize;
-            assert( !contains( dst[0 .. len], val ) );
+
             debug(info) printf( "appending (%.*s)\n", cast(int) val.length, val.ptr );
 
-            if ( dst.length - len >= val.length && &dst[len] == &val[0] )
-            {
-                // data is already in place
-                auto t = dst[len .. len + val.length];
-                len += val.length;
-                return t;
-            }
-            if ( dst.length - len >= val.length )
-            {
+            checkAndStretchBuf(val.length);
+
+            // data is already not in place?
+            if ( &dst[len] != &val[0] )
                 dst[len .. len + val.length] = val[];
-                auto t = dst[len .. len + val.length];
-                len += val.length;
-                return t;
-            }
-            overflow();
+
+            len += val.length;
+        }
+    }
+
+    @nogc:
+
+    // from index to end of current buf
+    private scope bslice(size_t from) nothrow
+    {
+        return bslice(from, len);
+    }
+
+    private scope bslice(size_t from, size_t to) nothrow
+    {
+        return BufSlice(dst, from, to);
+    }
+
+    private scope bslice_empty() nothrow
+    {
+        return BufSlice(null, 0, 0);
+    }
+}
+
+private struct BufSlice
+{
+    char[] dst;
+
+    @safe:
+    pure:
+    nothrow:
+
+    @disable this();
+
+    this(return scope char[] dst) scope nothrow @nogc
+    {
+        this(dst, 0, 0);
+    }
+
+    this(return scope char[] dst, size_t from, size_t to, bool lastArgIsLen = false) scope nothrow @nogc
+    {
+        this.dst = dst;
+        this.from = from;
+
+        if (lastArgIsLen)
+            this.to = from + to;
+        else
+            this.to = to;
+    }
+
+    size_t from;
+    size_t to;
+
+    invariant()
+    {
+        if (dst is null)
+        {
+            assert(from == 0);
+            assert(to == 0);
         }
-        return null;
+
+        assert(from <= to);
     }
+
+    auto getSlice() inout nothrow scope { return dst[from .. to]; }
```

Comment:
`opSlice` would have been more idiomatic here.

Reply:
I decided not, because it returns not slice of slice, but slice from dst array. It will be confusing